src/App.tsx

```diff
@@ -1,24 +1,283 @@
+    );
+  };
+
+  const handleSubmit = async (
+    event: FormEvent<HTMLFormElement>,
+    newTodoText: string,
+    setNewTodoText: Dispatch<SetStateAction<string>>,
+  ) => {
+    event.preventDefault();
+
+    if (!newTodoText.trim()) {
+      onError(ErrorType.EmptyTitle);
+
+      return;
+    }
+
+    const newTodo = {
+      id: 0,
+      userId: USER_ID,
+      title: newTodoText.trim(),
+      completed: false,
+    };
+
+    try {
+      setTempTodo(newTodo);
+      const todo = await addTodo(newTodo);
+
+      setTodosFromServer(
+        todos => ([...todos, todo]),
+      );
+
+      setNewTodoText('');
+    } catch {
+      onError(ErrorType.UnableToAdd);
+    } finally {
+      setTempTodo(() => null);
+
+      setTimeout(() => {
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+      }, 100);
+    }
+  };
+
+  const handleSingleTodoDelition = async (todoId: number) => {
+    try {
+      setSelectedTodoIds(ids => [...ids, todoId]);
+
+      await deleteTodo(todoId);
+
+      setTodosFromServer(
+        todos => todos.filter(todo => todo.id !== todoId),
+      );
+    } catch {
+      onError(ErrorType.UnableToDelete);
+    } finally {
+      setTimeout(
+        () => {
+          setSelectedTodoIds(
+            ids => ids.filter(id => !(id === todoId)),
+          );
+
+          if (inputRef.current) {
+            inputRef.current.focus();
+          }
+        },
+        300,
+      );
+    }
+  };
+
+  const completedTodoIds = todosFromServer
+    .filter(todo => todo.completed)
+    .map(todo => todo.id);
+
+  const handleCompletedTodosDelition = async () => {
+    const processedIds: number[] = [];
+
+    try {
+      setSelectedTodoIds(ids => [...ids, ...completedTodoIds]);
+
+      completedTodoIds.forEach(async (todoId) => {
+        try {
+          await deleteTodo(todoId);
+          processedIds.push(todoId);
+        } catch {
+          onError(ErrorType.UnableToDelete);
+        }
+      });
+    } catch {
+      onError(ErrorType.UnableToDelete);
+    } finally {
+      setTimeout(
+        () => {
+          setSelectedTodoIds(
+            ids => ids.filter(id => !completedTodoIds.includes(id)),
+          );
+
+          setTodosFromServer(
+            todos => todos.filter(todo => !processedIds.includes(todo.id)),
+          );
+
+          if (inputRef.current) {
+            inputRef.current.focus();
+          }
+        },
+        300,
+      );
+    }
+  };
+
+  const updateTodo = async (updatedTodo: Todo) => {
+    try {
+      setSelectedTodoIds(ids => [...ids, updatedTodo.id]);
+
+      const todoFromResponse = await patchTodo(updatedTodo);
+
+      setTodosFromServer(
+        todos => todos.map(td => {
+          if (td.id === todoFromResponse.id) {
+            return todoFromResponse;
+          }
+
+          return td;
+        }),
+      );
+    } catch {
+      onError(ErrorType.UnableToUpdate);
+      throw new Error(ErrorType.UnableToUpdate);
```

Comment:
You don't need to throw an error as it's already shown for the user

Reply:
The error throw is needed as the function is being reused in another handler that needs to catch an error (`onSumbitTitle` in `TodoItem.tsx`) ![KEKWait (2)](https://github.com/mate-academy/react_todo-app-with-api/assets/146858277/28d61293-b8d6-45ad-a8c2-f563f9ba4e0d)